it/lessons/advanced/concurrency.md

```diff
@@ -0,0 +1,175 @@
+  end
+end
+
+iex> Example.add(2, 3)
+5
+:ok
+```
+
+Per lanciare la stessa funzione in modo asincrono, usiamo `spawn/3`:
+
+```elixir
+iex> spawn(Example, :add, [2, 3])
+5
+#PID<0.80.0>
+```
+
+### Message Passing
+
+Per comunicare, i processi usano _message passing_. A riguardo, ritroviamo due componenti principali: `send/2` e `receive/1`. La funzione `send/2` ci permette di mandare message ad altri PID; per ascoltare per messaggi in arrivo, usiamo `receive/1` ed (opzionalmente) eseguiamo un _match_ sui messaggi in arrivo. Se un messaggio non soddisfa nessun _match_, il processo ignora il messaggio e continua ininterrotto.
+
+```elixir
+defmodule Example do
+  def listen do
+    receive do
+      {:ok, "hello"} -> IO.puts "World"
+    end
+  end
+end
+
+iex> pid = spawn(Example, :listen, [])
+#PID<0.108.0>
+
+iex> send pid, {:ok, "hello"}
+World
+{:ok, "hello"}
+
+iex> send pid, :ok
+:ok
+```
+
+### Process Linking
+
+Un problema con `spawn` è sapere quando il processo lanciato muore. Per questo motivo, abbiamo bisogno di collegare i nostri processi, usando `spawn_link`. Due processi collegati riceveranno notifiche d'uscita vicendevolmente.
+
+```elixir
+defmodule Example do
+  def explode, do: exit(:kaboom)
+end
+
+iex> spawn(Example, :explode, [])
+#PID<0.66.0>
+
+iex> spawn_link(Example, :explode, [])
+** (EXIT from #PID<0.57.0>) :kaboom
+```
+
+A volte, non vogliamo che un processo collegato uccida il processo che l'ha creato. Per evitare questo, dobbiamo _intrappolare le uscite_ nel processo genitore. Quando intrappoliamo uscite in un processo, questo riceverà solo messaggi di notifica in questa forma: `{:EXIT, dal_pid, ragione}`.
+
+```elixir
+defmodule Example do
+  def explode, do: exit(:kaboom)
+  def run do
+    Process.flag(:trap_exit, true)
+    spawn_link(Example, :explode, [])
+
+    receive do
+      {:EXIT, from_pid, reason} -> IO.puts "Exit reason: #{reason}"
+    end
+  end
+end
+
+iex> Example.run
+Exit reason: kaboom
+:ok
+```
+
+### Process Monitoring
+
+Cosa facciamo quando non vogliamo che due processi siano collegati, ma rimangano comunque informati sull'altro? In questo caso, possiamo fare _process monitoring_ con `spawn_monitor`. Quando monitoriamo un processo, riceviamo un messaggio quando questo muore, senza preoccuparci di ripercussioni sul processo corrente, e senza dover intrappolare uscite.
+
+```elixir
+defmodule Example do
+  def explode, do: exit(:kaboom)
+  def run do
+    {pid, ref} = spawn_monitor(Example, :explode, [])
+
+    receive do
+      {:DOWN, ref, :process, from_pid, reason} -> IO.puts "Exit reason: #{reason}"
+    end
+  end
+end
+
+iex> Example.run
+Exit reason: kaboom
+:ok
+```
+
+## Agenti
+
+Gli Agenti sono astrazioni attorno a dei processi di background, che mantengono uno stato interno. Possiamo accedere ad agenti da altri processi all'interno della nostra applicazione (e nodo). Lo stato di un agente è determinato dal valore di ritorno della funzione che lo inizializza, o aggiorna:
```

Comment:
Gli Agenti sono un'astrazione che permette di accedere e manipolare in maniera agevole delle informazioni riguardanti lo stato di una parte di applicazione. Possiamo accedere ad agenti da altri processi all'interno della nostra applicazione (e nodo). Lo stato di un agente è determinato dal valore di ritorno della funzione che lo inizializza o lo aggiorna: